File: app/services/search_service.py
"""Global Search module (PRD section 14)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import List

from app.core.database import get_db


@dataclass
class SearchResult:
    category: str  # "Child" | "Subject" | "Teacher" | "Academic Year" | "School" | "Expense Category"
    label: str
    detail: str
    ref_id: int = None


def search_all(query: str, conn=None) -> List[SearchResult]:
    query = query.strip()
    if not query:
        return []
    conn = conn or get_db()
    like = f"%{query}%"
    results: List[SearchResult] = []

    for row in conn.execute(
        "SELECT id, full_name, current_class, school_name FROM children WHERE full_name LIKE ?", (like,)
    ):
        results.append(SearchResult(
            "Child", row["full_name"],
            f"{row['current_class'] or ''} · {row['school_name'] or ''}".strip(" ·"),
            row["id"],
        ))

    for row in conn.execute("SELECT id, name FROM subjects WHERE name LIKE ?", (like,)):
        results.append(SearchResult("Subject", row["name"], "", row["id"]))

    for row in conn.execute(
        "SELECT id, name, subject, school_name FROM teachers WHERE name LIKE ?", (like,)
    ):
        results.append(SearchResult(
            "Teacher", row["name"], f"{row['subject'] or ''} · {row['school_name'] or ''}".strip(" ·"), row["id"]
        ))

    for row in conn.execute(
        "SELECT DISTINCT year_label FROM academic_years WHERE year_label LIKE ?", (like,)
    ):
        results.append(SearchResult("Academic Year", row["year_label"], ""))

    for row in conn.execute("SELECT id, name FROM schools WHERE name LIKE ?", (like,)):
        results.append(SearchResult("School", row["name"], "", row["id"]))

    for row in conn.execute(
        "SELECT DISTINCT school_name FROM children WHERE school_name LIKE ?", (like,)
    ):
        if row["school_name"]:
            results.append(SearchResult("School", row["school_name"], ""))

    for row in conn.execute(
        "SELECT id, name FROM expense_categories WHERE name LIKE ?", (like,)
    ):
        results.append(SearchResult("Expense Category", row["name"], "", row["id"]))

    return results

File: android-native/app/src/main/java/org/familytools/educationtracker/services/SearchService.kt
package org.familytools.educationtracker.services

import android.content.Context
import org.familytools.educationtracker.data.AppDatabase

data class SearchResult(val category: String, val label: String, val detail: String, val refId: Long? = null)

object SearchService {
    fun search(context: Context, query: String): List<SearchResult> {
        if (query.isBlank()) return emptyList()
        val db = AppDatabase.getInstance(context).openHelper.readableDatabase
        val like = "%$query%"
        val results = mutableListOf<SearchResult>()

        db.query("SELECT id, fullName, currentClass, schoolName FROM children WHERE fullName LIKE ?", arrayOf(like)).use {
            while (it.moveToNext()) {
                results.add(
                    SearchResult(
                        "Child", it.getString(1),
                        listOfNotNull(it.getString(2).ifBlank { null }, it.getString(3).ifBlank { null }).joinToString(" · "),
                        it.getLong(0),
                    ),
                )
            }
        }
        db.query("SELECT id, name FROM subjects WHERE name LIKE ?", arrayOf(like)).use {
            while (it.moveToNext()) results.add(SearchResult("Subject", it.getString(1), "", it.getLong(0)))
        }
        db.query("SELECT id, name, subject, schoolName FROM teachers WHERE name LIKE ?", arrayOf(like)).use {
            while (it.moveToNext()) {
                results.add(
                    SearchResult(
                        "Teacher", it.getString(1),
                        listOfNotNull(it.getString(2).ifBlank { null }, it.getString(3).ifBlank { null }).joinToString(" · "),
                        it.getLong(0),
                    ),
                )
            }
        }
        db.query("SELECT DISTINCT yearLabel FROM academic_years WHERE yearLabel LIKE ?", arrayOf(like)).use {
            while (it.moveToNext()) results.add(SearchResult("Academic Year", it.getString(0), ""))
        }
        db.query("SELECT DISTINCT schoolName FROM children WHERE schoolName LIKE ? AND schoolName != ''", arrayOf(like)).use {
            while (it.moveToNext()) results.add(SearchResult("School", it.getString(0), ""))
        }
        db.query("SELECT id, name FROM expense_categories WHERE name LIKE ?", arrayOf(like)).use {
            while (it.moveToNext()) results.add(SearchResult("Expense Category", it.getString(1), "", it.getLong(0)))
        }
        return results
    }
}
